### PR TITLE
Builders for partition filter

### PR DESCRIPTION
**Pull request by maysamyabandeh**
This is the second split of this pull request: https://github.com/facebook/rocksdb/pull/1891 which includes only the builder part. The testing will be included in the third split, where the reader is also included.

### CODE REVIEW

File: table/filter_block.h [hunk 1]
```diff
@@ -51,7 +51,12 @@ class FilterBlockBuilder {
   virtual bool IsBlockBased() = 0;                    // If is blockbased filter
   virtual void StartBlock(uint64_t block_offset) = 0;  // Start new block filter
   virtual void Add(const Slice& key) = 0;      // Add a key to current filter
-  virtual Slice Finish() = 0;                     // Generate Filter
+  Slice Finish() {                             // Generate Filter
+    const BlockHandle empty_handle;
+    Status dont_care_status;
+    return Finish(empty_handle, &dont_care_status);
```
Comment (yiwu-arbug) [inline table/filter_block.h]: assert(status.ok()) ? Just to avoid someone misuse it..

File: table/full_filter_block.cc [hunk 2]
```diff
@@ -40,11 +40,12 @@ inline void FullFilterBlockBuilder::AddKey(const Slice& key) {
 // Add prefix to filter if needed
 inline void FullFilterBlockBuilder::AddPrefix(const Slice& key) {
   Slice prefix = prefix_extractor_->Transform(key);
-  filter_bits_builder_->AddKey(prefix);
-  num_added_++;
```
Comment (yiwu-arbug) [inline table/full_filter_block.cc]: why this change?

File: table/filter_block.h (hunk 1 above)
Comment (maysamyabandeh) [inline table/filter_block.h]: sure. good thinking.

File: table/partitioned_filter_block.h [hunk 3]
```diff
@@ -7,59 +7,47 @@
...
 #include <string>
+#include <vector>
+#include "db/dbformat.h"
 #include "rocksdb/options.h"
 #include "rocksdb/slice.h"
 #include "rocksdb/slice_transform.h"
 #include "util/hash.h"
 
+#include "table/full_filter_block.h"
 #include "table/index_builder.h"
```
Comment (yiwu-arbug) [inline table/partitioned_filter_block.h]: nit: cleanup unused headers?

File: table/partitioned_filter_block.h [hunk 4]
```diff
@@ -7,59 +7,47 @@
...
+class PartitionedFilterBlockBuilder : public FullFilterBlockBuilder {
  public:
-  explicit PartitionIndexBuilder(const InternalKeyComparator* comparator,
-                                 const SliceTransform* prefix_extractor,
-                                 const uint64_t index_per_partition,
-                                 int index_block_restart_interval);
+  explicit PartitionedFilterBlockBuilder(
+      const SliceTransform* prefix_extractor, bool whole_key_filtering,
+      FilterBitsBuilder* filter_bits_builder, int index_block_restart_interval,
+      PartitionedIndexBuilder* const part_index_builder);
```
Comment (yiwu-arbug) [inline table/partitioned_filter_block.h]: I would prefer just `p_index_builder`.

File: table/index_builder.cc [hunk 5]
```diff
@@ -49,4 +51,100 @@ IndexBuilder* IndexBuilder::CreateIndexBuilder(
...
+PartitionedIndexBuilder* PartitionedIndexBuilder::CreateIndexBuilder(
+    const InternalKeyComparator* comparator,
+    const SliceTransform* prefix_extractor, int index_block_restart_interval,
+    uint64_t index_per_partition, const BlockBasedTableOptions& table_opt) {
+  class DummyFilterBitsBuilder : public FilterBitsBuilder {
+   public:
+    virtual void AddKey(const Slice& key) { assert(0); }
+    virtual Slice Finish(std::unique_ptr<const char[]>* buf) { assert(0); }
+  };
+  FilterBitsBuilder* filter_bits_builder =
```
Comment (yiwu-arbug) [inline table/index_builder.cc]: unused?

File: table/partitioned_filter_block.h (hunk 3 above)
Comment (yiwu-arbug) [inline table/partitioned_filter_block.h]: cleanup unused headers?

File: table/partitioned_filter_block.cc [hunk 6]
```diff
@@ -9,76 +9,61 @@
...
-    sub_index_builder_ =
-        CreateIndexBuilder(sub_type_, comparator_, prefix_extractor_,
-                           index_block_restart_interval_, index_per_partition_);
+// FullFilterBlockBuilder
+void PartitionedFilterBlockBuilder::MayBeCutAFilterBlock() {
+  if (!part_index_builder_->ShouldCutFilterBlock()) {
+    return;
   }
+  filter_gc.push_back(std::unique_ptr<const char[]>(nullptr));
+  Slice filter = filter_bits_builder_->Finish(&filter_gc.back());
```
Comment (yiwu-arbug) [inline table/partitioned_filter_block.cc]: Not familiar with filter bits builder. Will it restart after call finish?

File: table/partitioned_filter_block.cc [hunk 7]
```diff
@@ -9,76 +9,61 @@
...
+    // This is the full key vs. the user key in the filters. We assume that
+    // user key <= full key
+    index_on_filter_block_builder_.Add(last_entry.key, handle_encoding);
+    filters.pop_front();
   }
-  // If there is no sub_index left, then return the 2nd level index.
-  if (UNLIKELY(entries_.empty())) {
-    index_blocks->index_block_contents = index_block_builder_.Finish();
-    return Status::OK();
+  MayBeCutAFilterBlock();
```
Comment (yiwu-arbug) [inline table/partitioned_filter_block.cc]: seems you shouldn't call `MayBeCutAFitlerBlock` if `finishing_filters=true`?

File: table/partitioned_filter_block.h [hunk 8]
```diff
@@ -7,59 +7,47 @@
...
-  int index_block_restart_interval_;
-  uint64_t num_indexes = 0;
-  bool finishing =
+  std::list<FilterEntry> filters;  // list of partitioned indexes and their keys
+  std::unique_ptr<IndexBuilder> value;
+  std::vector<std::unique_ptr<const char[]>> filter_gc;
+  bool finishing_filters =
       false;  // true if Finish is called once but not complete yet.
+  // The policy of when cut a filter block and Finish it
+  void MayBeCutAFilterBlock();
```
Comment (yiwu-arbug) [inline table/partitioned_filter_block.h]: nit: s/MayBe/Maybe/

File: table/block_based_table_builder.cc [hunk 9]
```diff
@@ -60,19 +61,28 @@ typedef BlockBasedTableOptions::IndexType IndexType;
 // Without anonymous namespace here, we fail the warning -Wmissing-prototypes
 namespace {
 
-// Create a index builder based on its type.
-FilterBlockBuilder* CreateFilterBlockBuilder(const ImmutableCFOptions& opt,
-    const BlockBasedTableOptions& table_opt) {
+// Create a filter block builder based on its type.
+FilterBlockBuilder* CreateFilterBlockBuilder(
+    const ImmutableCFOptions& opt, const BlockBasedTableOptions& table_opt,
+    PartitionedIndexBuilder* const part_index_builder) {
```
Comment (yiwu-arbug) [inline table/block_based_table_builder.cc]: I would prefer s/part_index_builder/p_index_builder/

File: table/block_based_table_builder.cc [hunk 10]
```diff
@@ -277,21 +287,36 @@ struct BlockBasedTableBuilder::Rep {
...
+      index_builder.reset(part_index_builder);
+    } else {
+      index_builder.reset(IndexBuilder::CreateIndexBuilder(
+          table_options.index_type, &internal_comparator,
+          &this->internal_prefix_transform, _ioptions.prefix_extractor,
+          table_options.index_block_restart_interval,
+          table_options.index_per_partition, table_options));
+    }
+    if (skip_filters) {
+      filter_block = nullptr;
```
Comment (yiwu-arbug) [inline table/block_based_table_builder.cc]: s/filter_block/filter_builder/ ?

File: table/block_based_table_builder.cc [hunk 11]
```diff
@@ -374,6 +399,8 @@ void BlockBasedTableBuilder::Add(const Slice& key, const Slice& value) {
       }
     }
 
+    // Note: PartitionedFilterBlockBuilder requires adding to filter_block to be
```
Comment (yiwu-arbug) [inline table/block_based_table_builder.cc]: ... requires key being added to filter builder after being added to index builder.

File: table/index_builder.cc [hunk 12]
```diff
@@ -24,21 +25,22 @@ namespace rocksdb {
 IndexBuilder* IndexBuilder::CreateIndexBuilder(
     BlockBasedTableOptions::IndexType index_type,
     const InternalKeyComparator* comparator,
+    const SliceTransform* slice_transform,
```
Comment (maysamyabandeh) [inline table/index_builder.cc]: prefix_extractor comes from _ioptions.prefix_extractor which is of type SliceTransform.

slice_transform comes from this->internal_prefix_transform which is of type InternalKeySliceTransform initialized by prefix_extractor's transform member. slice_transform is used only by the hash index builder.

previously I was reusing InternalKeySliceTransform for also partitioned filter builder which turns out to be wrong. So here we distinguish between the two, use slice_transform for hash (as it was before) and use _ioptions.prefix_extractor for partitioned filter.

I understand the naming is confusing here. I am going to ahead doing this renaming:
slice_transform => int_key_slice_transform
prefox_extractor => slice_transform and also use the explicit type of InternalKeySliceTransform for it.

File: table/index_builder.cc [hunk 13]
```diff
@@ -24,21 +25,22 @@ namespace rocksdb {
...
+      return new HashIndexBuilder(comparator, slice_transform,
                                   index_block_restart_interval);
     }
     case BlockBasedTableOptions::kTwoLevelIndexSearch: {
-      return new PartitionIndexBuilder(comparator, prefix_extractor,
-                                       index_per_partition,
-                                       index_block_restart_interval);
+      return PartitionedIndexBuilder::CreateIndexBuilder(
+          comparator, prefix_extractor, index_block_restart_interval,
+          index_per_partition, table_opt);
```
Comment (maysamyabandeh) [inline table/index_builder.cc]: yeah i had to add table_opt later in the coding. yeah it makes sense to remove the two parameters.

File: table/index_builder.cc (hunk 5 above)
Comment (maysamyabandeh) [inline table/index_builder.cc]: good catch. it was needed when index and filter builders were mixed. but not anymore.

File: table/index_builder.h [hunk 14]
```diff
@@ -262,4 +265,72 @@ class HashIndexBuilder : public IndexBuilder {
...
+  int index_block_restart_interval_;
+  uint64_t num_indexes = 0;
+  bool finishing_indexes =
+      false;  // true if Finish is called once but not complete yet.
+  const BlockBasedTableOptions& table_opt_;
+  // Filter data
+  bool cut_filter_block =
+      false;  // true if it should cut the next filter partition block
+};
+}
```
Comment (maysamyabandeh) [inline table/index_builder.h]: why is that "make format" does not report this issues?

File: table/partitioned_filter_block.h (hunk 3 above)
Comment (maysamyabandeh) [inline table/partitioned_filter_block.h]: yeah it should be block_builder.h

File: table/partitioned_filter_block.h (hunk 4 above)
Comment (maysamyabandeh) [inline table/partitioned_filter_block.h]: ok

File: table/partitioned_filter_block.h (hunk 8 above)
Comment (maysamyabandeh) [inline table/partitioned_filter_block.h]: sure

File: table/partitioned_filter_block.cc (hunk 6 above)
Comment (maysamyabandeh) [inline table/partitioned_filter_block.cc]: yeah: https://github.com/facebook/rocksdb/blob/master/util/bloom.cc#L72

File: table/partitioned_filter_block.cc (hunk 7 above)
Comment (maysamyabandeh) [inline table/partitioned_filter_block.cc]: when `finishing_filters=true` then MayBeCutAFitlerBlock will be essentially noop. but It would good to make it clear in the code.

File: table/block_based_table_builder.cc (hunk 9 above)
Comment (maysamyabandeh) [inline table/block_based_table_builder.cc]: ok

File: table/block_based_table_builder.cc (hunk 10 above)
Comment (maysamyabandeh) [inline table/block_based_table_builder.cc]: oh yeah you said before. sure.

File: table/block_based_table_builder.cc (hunk 11 above)
Comment (maysamyabandeh) [inline table/block_based_table_builder.cc]: sure.

File: table/index_builder.cc (hunk 12 above)
Comment (maysamyabandeh) [inline table/index_builder.cc]: Actually I realized that after separating index and filter builders prefix_extractor is not need anymore, so i am going to remove it too.

File: table/partitioned_filter_block.h (hunk 3 above)
Comment (maysamyabandeh) [inline table/partitioned_filter_block.h]: that is actually is not unused. it is needed for partitionedindexbuilder

File: table/index_builder.h (hunk 14 above)
Comment (yiwu-arbug) [inline table/index_builder.h]: `make format` probably have different set of lint rules than what phabricator have..

File: table/partitioned_filter_block.h (hunk 3 above)
Comment (yiwu-arbug) [inline table/partitioned_filter_block.h]: I mean all the headers list here. Is any of them unused? I didn't check though.

File: table/partitioned_filter_block.cc [hunk 15]
```diff
@@ -13,27 +13,27 @@ namespace rocksdb {
 PartitionedFilterBlockBuilder::PartitionedFilterBlockBuilder(
     const SliceTransform* prefix_extractor, bool whole_key_filtering,
     FilterBitsBuilder* filter_bits_builder, int index_block_restart_interval,
-    PartitionedIndexBuilder* const part_index_builder)
+    PartitionedIndexBuilder* const p_index_builder)
```
Comment (yiwu-arbug) [inline table/partitioned_filter_block.cc]: This is a constant pointer. Are you wanting to have a pointer to constant (e.g. const T*) ?

Comment (maysamyabandeh) [inline table/partitioned_filter_block.cc]: No I meant a constant pointer as there is no reason for the pointer to be changed.

File: table/partitioned_filter_block.cc [hunk 16]
```diff
@@ -9,76 +9,62 @@
...
-Status PartitionIndexBuilder::Finish(
-    IndexBlocks* index_blocks, const BlockHandle& last_partition_block_handle) {
-  assert(!entries_.empty());
-  // It must be set to null after last key is added
-  assert(sub_index_builder_ == nullptr);
-  if (finishing == true) {
-    Entry& last_entry = entries_.front();
+Slice PartitionedFilterBlockBuilder::Finish(
+    const BlockHandle& last_partition_block_handle, Status* status) {
+  if (finishing_filters == true) {
```
Comment (yiwu-arbug) [inline table/partitioned_filter_block.cc]: It took me some time to verify the logic of this method. Would this be a little bit more clear?
```
if (filter.empty() && !finishing_filters) {
  return Slice();
} else if (!filter.empty()) {
  auto filter = filters.front();
  filters.pop_front();
  finish_filters = true;
  index_builder.Add(filter.key);
  return filter.filter;
} else {
  return index_builder.Finish();
}
